contract/contracts/forward_name_resolver/forward_name_resolver.ral

```diff
@@ -0,0 +1,259 @@
+    }
+
+    @using(preapprovedAssets = true, checkExternalCaller = false)
+    pub fn generateToken() -> ByteVec {
+        let tokenExists = contractExists!(subContractId!(getRewardTokenKey()))
+        assert!(!tokenExists, ErrorCodes.TokenAlreadyGenerated)
+        let caller = callerAddress!()
+
+        let (encodeImmutableFields, encodeMutableFields) = RewardToken.encodeFields!(
+            selfContractId!(),
+            b`ANS`,
+            b`Alephium Name Service`,
+            18,
+            0
+        )
+
+        return copyCreateSubContractWithToken!{caller -> ALPH: minimalContractDeposit!()}(
+            getRewardTokenKey(),
+            tokenTemplateId,
+            encodeImmutableFields,
+            encodeMutableFields,
+            u256Max!()
+        )
+    }
+
+    @using(preapprovedAssets = true)
+    fn createName(minter: Address, index: U256, name: ByteVec, capitalisation: ByteVec) -> ByteVec {
+        let now = blockTimeStamp!()
+        let expires = now + renewLength
+        let (encodeImmutableFields, encodeMutableFields) = Name.encodeFields!(
+            selfContractId!(),  // <- collectionId
+            index,              // <- nftIndex
+            name,               // <- name
+            capitalisation,     // <- capitalisation
+            minter,             // <- initialAddress
+            expires             // <- expires
+        )
+
+        let newContractId = copyCreateSubContractWithToken!{minter -> ALPH: minimalContractDeposit!()}(
+            getNftKey(index),
+            nameTemplateId,
+            encodeImmutableFields,
+            encodeMutableFields,
+            1,
+            minter
+        )
+
+        emit NameCreated(
+            index,
+            name,
+            capitalisation,
+            minter,
+            expires
+        )
+
+        return newContractId
+    }
+
+    @using(preapprovedAssets = true)
+    fn handleRewardTokens(caller: Address, name: ByteVec) -> () {
+        let nameLength = size!(name)
+        let rewardToken = getRewardToken()
+        let rewardTokenId = tokenId!(rewardToken)
+
+        if (nameLength > 8) {
+            rewardToken.mintTokens(caller, 1 alph)
+        }
+        if (nameLength <= 8) {
+            let cost = (9 - nameLength) * 100
+            rewardToken.burnTokens{caller -> rewardTokenId: cost * 1 alph}(caller, cost * 1 alph)
+        }
+    }
+
+    @using(preapprovedAssets = true, updateFields = true, checkExternalCaller = false)
+    pub fn mint(name: ByteVec, capitalisation: ByteVec) -> ByteVec {
+        let caller = callerAddress!()
+        totalSupply = totalSupply + 1
+        let rewardToken = getRewardToken()
+        let rewardTokenId = tokenId!(rewardToken)
+        
+        if (nameNftIndex.contains!(name)) {
+            let existingNft = getNftByName(name)
+            let expired = hasExpired(existingNft)
+            assert!(expired, ErrorCodes.NameHasNotExpired)
+            existingNft.delete(caller)
```

Comment:
Should it be `existingNft.delete(previousNameOwner)` instead of the `existingNft.delete(caller)`? Otherwise, the ALPH deposited for the name NFT will be transferred to the caller instead of the previous name owner.

Reply:
We don't actually know the previous owner, I wasn't sure if it matters who gets the storage fee.

Reply:
I am thinking that perhaps we could store the index and the minter address in `nameNftIndex`. This way, when destroying, the deposited ALPH can be refunded to the correct address. But I am not sure if it is worth doing so.

Reply:
Often people will sell the NFT anyway so they don't really deserve the deposit back